File: book2-algorithms-sedgewick/ch1/ex-1.2.6-circular-shift.py
# Algorithms: Sedgewick, Ch1 Ex 1.2.6 Circular Shift


def is_circular_rotation(s1: str, s2: str) -> bool:
    """
    Returns true if s2 is a circular rotation of s1.
    """
    return s2 in s1 + s1


def run():
    test_cases = [
        ("", "", True),
        ("A", "A", True),
        ("AA", "AA", True),
        ("ABC", "CAB", True),
        ("ABA", "BAA", True),
        ("ABA", "AAC", False),
        ("ACTGACG", "TGACGAC", True),
        ("CTGACG", "TGACGAC", False),
    ]
    for test_case in test_cases:
        print(
            f"is_circular_rotation({test_case[0]}, {test_case[1]}): ",
            is_circular_rotation(test_case[0], test_case[1]),
            f"Expected: {test_case[2]}",
        )


if __name__ == "__main__":
    run()

File: book2-algorithms-sedgewick/ch1/s01_exercises_rust/src/ex_1_2_6_circular_shift.rs
fn is_circular_rotation(s1: &str, s2: &str) -> bool {
    let s3 = String::from(s1) + s1;
    s3.find(s2).is_some()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_is_circular_rotation() {
        let test_cases: [(&str, &str, bool); 8] = [
            ("", "", true),
            ("A", "A", true),
            ("AA", "AA", true),
            ("ABC", "CAB", true),
            ("ABA", "BAA", true),
            ("ABA", "AAC", false),
            ("ACTGACG", "TGACGAC", true),
            ("CTGACG", "TGACGAC", false),
        ];
        for (s1, s2, expected) in test_cases {
            assert_eq!(is_circular_rotation(s1, s2), expected);
        }
    }
}
